Provider crashes while trying to read policy

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform AWS Provider Version

Terraform v1.2.6
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v4.23.0
+ provider registry.terraform.io/hashicorp/random v3.3.2

### Affected Resource(s)


### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
data "aws_iam_roles" "platformoperatorepcloudteam" {
    name_regex = "AWSReservedSSO_platformoperatorepcloudteam_.*"
}

data "aws_iam_policy_document" "kms_key_base" {
  statement {
    effect = "Allow"
    actions = [
      "kms:Create*",
      "kms:Describe*",
      "kms:Enable*",
      "kms:List*",
      "kms:Put*",
      "kms:Update*",
      "kms:Revoke*",
      "kms:Disable*",
      "kms:Get*",
      "kms:Delete*",
      "kms:ScheduleKeyDeletion",
      "kms:CancelKeyDeletion",
      "kms:Tag*",
      "kms:UntagResource"
    ]
    principals {
      type        = "AWS"
      identifiers = ["arn:aws:iam::${local.account_id}:root"]
    }
    resources = ["*"]
  }
  statement {
      effect = "Allow"
      actions = [
          "kms:Create*",
          "kms:Describe*",
          "kms:Enable*",
          "kms:List*",
          "kms:Put*",
          "kms:Update*",
          "kms:Revoke*",
          "kms:Disable*",
          "kms:Get*",
          "kms:Delete*",
          "kms:TagResource",
          "kms:UntagResource",
          "kms:ScheduleKeyDeletion",
          "kms:CancelKeyDeletion"
      ]
  principals {
          type        = "AWS"
          identifiers = data.aws_iam_roles.platformoperatorepcloudteam.arns 
          }
          resources = ["*"]
    
    }
}

data "aws_iam_policy_document" "merged_kms_policy" {
  source_policy_documents = [
    data.aws_iam_policy_document.kms_key_base.json,
    var.policy
  ]
}
```

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output

```
Stack trace from the terraform-provider-aws_v4.23.0_x5 plugin:

panic: interface conversion: interface {} is nil, not string

goroutine 572 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/iam.dataSourcePolicyDocumentRead(0x0, {0x0, 0x0})
	github.com/hashicorp/terraform-provider-aws/internal/service/iam/policy_document_data_source.go:140 +0x1194
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xab8eae0, {0xab8eae0, 0xc0020499b0}, 0xd, {0x84c7d60, 0xc000fe9500})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.19.0/helper/schema/resource.go:712 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).ReadDataApply(0xc0008d2fc0, {0xab8eae0, 0xc0020499b0}, 0xc002b36c00, {0x84c7d60, 0xc000fe9500})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.19.0/helper/schema/resource.go:943 +0x145
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadDataSource(0xc000292fc0, {0xab8eae0, 0xc0020497d0}, 0xc0025f70a0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.19.0/helper/schema/grpc_provider.go:1179 +0x408
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.muxServer.ReadDataSource({0xc0016e97d0, 0xc0016e9830, {0xc0007b53e0, 0x2, 0x2}, 0xc0016e9800, 0xc00119dc70, 0xc00296bcb0, 0xc0016e9860}, {0xab8eae0, ...}, ...)
	github.com/hashicorp/terraform-plugin-mux@v0.7.0/tf5muxserver/mux_server_ReadDataSource.go:27 +0x142
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadDataSource(0xc0019b6140, {0xab8eae0, 0xc002056420}, 0xc00275e320)
	github.com/hashicorp/terraform-plugin-go@v0.12.0/tfprotov5/tf5server/server.go:658 +0x445
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler({0x971ff00, 0xc0019b6140}, {0xab8eae0, 0xc002056420}, 0xc0011fcd20, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.12.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:421 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc002422380, {0xaca0a30, 0xc0007deb60}, 0xc002b83e60, 0xc00065e9c0, 0x10b7e3f0, 0x0)
	google.golang.org/grpc@v1.48.0/server.go:1295 +0xb03
google.golang.org/grpc.(*Server).handleStream(0xc002422380, {0xaca0a30, 0xc0007deb60}, 0xc002b83e60, 0x0)
	google.golang.org/grpc@v1.48.0/server.go:1636 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.48.0/server.go:932 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.48.0/server.go:930 +0x294

Error: The terraform-provider-aws_v4.23.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

Operation failed: failed running terraform plan (exit 1)
```
### Expected Behavior

a terraform run

### Actual Behavior

a crash

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform apply`

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->

* #0000

--- a/.changelog/26640.txt
+++ b/.changelog/26640.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+ data-source/aws_iam_policy_document: Prevent crash when `source_policy_documents` contains empty or invalid JSON documents
+```

--- a/internal/service/iam/policy_document_data_source.go
+++ b/internal/service/iam/policy_document_data_source.go
@@ -56,7 +56,7 @@ func DataSourcePolicyDocument() *schema.Resource {
 				Optional: true,
 				Elem: &schema.Schema{
 					Type:         schema.TypeString,
-					ValidateFunc: validation.All(validation.StringIsNotEmpty, validation.StringIsJSON),
+					ValidateFunc: validation.StringIsJSON,
 				},
 			},
 			"statement": {
@@ -139,6 +139,10 @@ func dataSourcePolicyDocumentRead(d *schema.ResourceData, meta interface{}) erro
 
 		// merge sourceDocs in order specified
 		for sourceJSONIndex, sourceJSON := range v.([]interface{}) {
+			if sourceJSON == nil {
+				continue
+			}
+
 			sourceDoc := &IAMPolicyDoc{}
 			if err := json.Unmarshal([]byte(sourceJSON.(string)), sourceDoc); err != nil {
 				return err

--- a/internal/service/iam/policy_document_data_source.go
+++ b/internal/service/iam/policy_document_data_source.go
@@ -54,7 +54,10 @@ func DataSourcePolicyDocument() *schema.Resource {
 			"source_policy_documents": {
 				Type:     schema.TypeList,
 				Optional: true,
-				Elem:     &schema.Schema{Type: schema.TypeString},
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validation.All(validation.StringIsNotEmpty, validation.StringIsJSON),
+				},
 			},
 			"statement": {
 				Type:     schema.TypeList,

--- a/internal/service/iam/policy_document_data_source_test.go
+++ b/internal/service/iam/policy_document_data_source_test.go
@@ -223,6 +223,24 @@ func TestAccIAMPolicyDocumentDataSource_duplicateSid(t *testing.T) {
 	})
 }
 
+func TestAccIAMPolicyDocumentDataSource_sourcePolicyValidJson(t *testing.T) {
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ErrorCheck:               acctest.ErrorCheck(t, iam.EndpointsID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccPolicyDocumentDataSourceConfig_invalidJson,
+				ExpectError: regexp.MustCompile(`"source_policy_documents.0" contains an invalid JSON: unexpected end of JSON input`),
+			},
+			{
+				Config:      testAccPolicyDocumentDataSourceConfig_emptyString,
+				ExpectError: regexp.MustCompile(`expected "source_policy_documents.0" to not be an empty string, got`),
+			},
+		},
+	})
+}
+
 // Reference: https://github.com/hashicorp/terraform-provider-aws/issues/10777
 func TestAccIAMPolicyDocumentDataSource_StatementPrincipalIdentifiers_stringAndSlice(t *testing.T) {
 	dataSourceName := "data.aws_iam_policy_document.test"
@@ -1274,6 +1292,18 @@ data "aws_iam_policy_document" "test" {
       type = "AWS"
     }
   }
+}
+`
+
+var testAccPolicyDocumentDataSourceConfig_emptyString = `
+data "aws_iam_policy_document" "test" {
+	source_policy_documents = [""]
+}
+`
+
+var testAccPolicyDocumentDataSourceConfig_invalidJson = `
+data "aws_iam_policy_document" "test" {
+	source_policy_documents = ["{"]
 }
 `
 

--- a/internal/service/iam/policy_document_data_source_test.go
+++ b/internal/service/iam/policy_document_data_source_test.go
@@ -223,19 +223,23 @@ func TestAccIAMPolicyDocumentDataSource_duplicateSid(t *testing.T) {
 	})
 }
 
-func TestAccIAMPolicyDocumentDataSource_sourcePolicyValidJson(t *testing.T) {
+func TestAccIAMPolicyDocumentDataSource_sourcePolicyValidJSON(t *testing.T) {
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:                 func() { acctest.PreCheck(t) },
 		ErrorCheck:               acctest.ErrorCheck(t, iam.EndpointsID),
 		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
 		Steps: []resource.TestStep{
 			{
-				Config:      testAccPolicyDocumentDataSourceConfig_invalidJson,
+				Config:      testAccPolicyDocumentDataSourceConfig_invalidJSON,
 				ExpectError: regexp.MustCompile(`"source_policy_documents.0" contains an invalid JSON: unexpected end of JSON input`),
 			},
 			{
-				Config:      testAccPolicyDocumentDataSourceConfig_emptyString,
-				ExpectError: regexp.MustCompile(`expected "source_policy_documents.0" to not be an empty string, got`),
+				Config: testAccPolicyDocumentDataSourceConfig_emptyString,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.aws_iam_policy_document.test", "json",
+						testAccPolicyDocumentExpectedJSONNoStatement,
+					),
+				),
 			},
 		},
 	})
@@ -1297,15 +1301,19 @@ data "aws_iam_policy_document" "test" {
 
 var testAccPolicyDocumentDataSourceConfig_emptyString = `
 data "aws_iam_policy_document" "test" {
-	source_policy_documents = [""]
+  source_policy_documents = [""]
 }
 `
 
-var testAccPolicyDocumentDataSourceConfig_invalidJson = `
+var testAccPolicyDocumentDataSourceConfig_invalidJSON = `
 data "aws_iam_policy_document" "test" {
-	source_policy_documents = ["{"]
+  source_policy_documents = ["{"]
 }
 `
+
+var testAccPolicyDocumentExpectedJSONNoStatement = `{
+  "Version": "2012-10-17"
+}`
 
 func testAccPolicyDocumentExpectedJSONStatementPrincipalIdentifiersMultiplePrincipals() string {
 	return fmt.Sprintf(`{

--- a/internal/service/iam/policy_model.go
+++ b/internal/service/iam/policy_model.go
@@ -13,7 +13,7 @@ const (
 type IAMPolicyDoc struct {
 	Version    string                `json:",omitempty"`
 	Id         string                `json:",omitempty"`
-	Statements []*IAMPolicyStatement `json:"Statement"`
+	Statements []*IAMPolicyStatement `json:"Statement,omitempty"`
 }
 
 type IAMPolicyStatement struct {

